### PR TITLE
[Sema] Fix an out-of-bounds crash when diagnosing bad conversion for a function with a parameter pack.

### PR DESCRIPTION
**Pull request by hokein**
This fixes a regression issue caused by 176981ac58d3e4beb02b9d110524e72be509375d.

When clang diagnoses bad conversion on a particular function parameter, we use the Index `I` to get the corresponding parameter from the primary function decl, the `I` is the index of function parameter after template instantations, so we can hit an out-of-bounds crash when accessing the `I`th function parameter if the function has pack parameters. This PR fixes it.

### CODE REVIEW

File: clang/lib/Sema/SemaOverload.cpp [hunk 1]
```diff
@@ -11298,8 +11298,14 @@ static void DiagnoseBadConversion(Sema &S, OverloadCandidate *Cand,
...
   QualType FromTy = Conv.Bad.getFromType();
   QualType ToTy = Conv.Bad.getToType();
-  SourceRange ToParamRange =
-      !isObjectArgument ? Fn->getParamDecl(I)->getSourceRange() : SourceRange();
+  SourceRange ToParamRange;
+  if (!isObjectArgument) {
+    if (I < Fn->getNumParams())
+      ToParamRange = Fn->getParamDecl(I)->getSourceRange();
+    else
+      // parameter pack case.
```
Comment (shafik) [inline clang/lib/Sema/SemaOverload.cpp:11306]: While this comment does the right thing™️ wrt to the `else` I would just use braces b/c it is not obvious and would probably give a lot of folks pause.

**Review by shafik — COMMENTED** [review]
Can you add some details to the summary. What was the original code doing wrong and the proposed new approach.

File: clang/lib/Sema/SemaOverload.cpp [hunk 2]
```diff
@@ -11298,8 +11298,14 @@ static void DiagnoseBadConversion(Sema &S, OverloadCandidate *Cand,
...
+      ToParamRange = Fn->parameters().back()->getSourceRange();
```
[the remark this answers is not in the document]
Comment (hokein) [reply clang/lib/Sema/SemaOverload.cpp]: thanks for the test cases. (I wonder if there is a case where we should point at the second pack, but I don't come up with one, I think using the first pack is probably good enough. Alternatively, we could use the full `getParametersSourceRange` range without finding a particular pack, but it provides less precise location information).

> Should we put that in a function in FunctionDecl?

Probably not worth, it seems heavy to add a method in FunctionDecl to only fix this regression.

File: clang/lib/Sema/SemaOverload.cpp (hunk 1 above)
Comment (hokein) [reply clang/lib/Sema/SemaOverload.cpp]: Done.

**Pull request closed without merging**